tiledb/sm/filesystem/vfs.cc

```diff
@@ -389,18 +382,95 @@ void VFS::move_dir(const URI& old_uri, const URI& new_uri) const {
   }
 }
 
-void VFS::copy_file(const URI& old_uri, const URI& new_uri) const {
+void VFS::chunked_buffer_io(const URI& src, const URI& dest) {
+  auto &src_fs{get_fs(src)}, &dest_fs{get_fs(dest)};
+  std::condition_variable cv;
+  std::mutex mtx;
+
+  // Deque which stores the buffers passed between the readers and writer.
+  ProducerConsumerQueue<std::variant<std::string, std::exception_ptr>>
+      buffer_queue;
+
+  // By default, the buffer size is 10 MB, unless the filesize is smaller.
```

Comment:
10MB is pretty large, considering that copying a large file would consume ten times that amount of memory (100MB), which can be unacceptable by some use cases.

We should lower the buffer size and/or the size of the queue.

Reply:
My design dictates that the maximum memory footprint of the buffer queue be 100mb. Smaller buffer size results in an increase in I/O calls, increasing latency at the write bottleneck. Similarly, reducing the number of buffers in the queue results in less read-ahead, and more waiting at the top of the reader's critical section. 
This is open for discussion, but the algorithm was intentionally designed as-such.